src/Autofac.Extensions.DependencyInjection/AutofacServiceProvider.cs

```diff
@@ -24,27 +24,68 @@
 // OTHER DEALINGS IN THE SOFTWARE.
 
 using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
 using Microsoft.Extensions.DependencyInjection;
 
 namespace Autofac.Extensions.DependencyInjection
 {
     internal class AutofacServiceProvider : IServiceProvider, ISupportRequiredService
     {
         private readonly IComponentContext _componentContext;
+        private List<OrderedServiceDescriptor> _originalDescriptors;
+        private HashSet<Type> _serviceTypes;
+        private MethodInfo _castMethodInfo;
 
-        public AutofacServiceProvider(IComponentContext componentContext)
+        public AutofacServiceProvider(IComponentContext componentContext, List<ServiceDescriptor> originalDescriptors)
         {
             _componentContext = componentContext;
+            _originalDescriptors = originalDescriptors.Select((d, i) => new OrderedServiceDescriptor
+            {
+                Index = i,
+                ImplementationInstance = d.ImplementationInstance,
+                ImplementationType = d.ImplementationType
+            }).ToList();
+            _serviceTypes = new HashSet<Type>(originalDescriptors.Select(d => d.ServiceType));
+            _castMethodInfo = typeof(Enumerable).GetTypeInfo().GetDeclaredMethod("Cast");
         }
 
         public object GetService(Type serviceType)
         {
-            return _componentContext.ResolveOptional(serviceType);
+            return GetOrdered(serviceType, _componentContext.ResolveOptional(serviceType));
```

Comment:
I've looked at the code for constructor injection, specifically `Microsoft.Extensions.Internal.ActivatorUtilities` and it does seem like constructor injection does in fact use `IServiceProvider.GetService()`? This also seems to be the case in practical tests, I do hit the breakpoints i expect when I do constructor injection of a `IEnumerable<T>` in a controller for example. Am I missing something else @halter73 ?